### PR TITLE
feat(table-permission): supports table level authority control

### PR DESCRIPTION
**Pull request by smallsheeeep**
#### What type of PR is this?
type-feature
module-table permission

#### What this PR does / why we need it:
This PR add table level authority control ability to colse #1585 .

#### Special notes for your reviewer:
This PR is created based on codes contributed by @ISADBA in #2163 . So please **do not** use squash merge to keep the commit history @yhilmare @yizhouxw .

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-migrate/src/main/resources/migrate/common/V_4_3_0_4__alter_iam_permission.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+--
+-- Add columns in `iam_permission` table for table-level permission management
+--
+alter table `iam_permission` add column `resource_type` varchar(64) default null comment 'Refer to ResourceType, optional values: ODC_DATABASE, ODC_TABLE, ODC_COLUMN, etc.';
+alter table `iam_permission` add column `resource_id` bigint(20) default null comment 'ID of the resource';
+alter table `iam_permission` add key `idx_iam_permission_resource_type_id` (resource_type, resource_id);
+update `iam_permission` set `resource_type` = 'ODC_DATABASE', `resource_id` = substring( `resource_identifier` from 14 ) where `resource_identifier` like 'ODC_DATABASE:%'  and `action` in ('query', 'change', 'export');
```
Comment (yizhouxw) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_3_0_4__alter_iam_permission.sql:7]: where condition may contains 'resource_type is not null'

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/TableController.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * @Version 1.0
+ */
+@RestController
+@RequestMapping("/api/v2/table")
+public class TableController {
+
+    @Autowired
+    private TableService service;
+
+    @ApiOperation(value = "list", notes = "List the latest tables by session id")
```
Comment (yizhouxw) [inline server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/TableController.java:47]: `@ApiOperation(value = "listTables", notes = "List table metadatas")`

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/TableController.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import io.swagger.annotations.ApiOperation;
+
+/**
+ *
+ * @Author: fenghao
+ * @Create 2024/3/19 11:13
+ * @Version 1.0
+ */
+@RestController
+@RequestMapping("/api/v2/table")
```
Comment (yizhouxw) [inline server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/TableController.java:41]: `/api/v2/database-schema`

let's module it as 'database schema',  we may add `views`, `procedures`, `triggers` in future.

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/TablePermissionController.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    @ApiOperation(value = "batchCreateTablePermissions", notes = "Batch create table permissions")
+    @RequestMapping(value = "/batchCreate", method = RequestMethod.POST)
+    public ListResponse<UserTablePermission> batchCreate(@PathVariable Long projectId,
+            @RequestBody CreateTablePermissionReq req) {
+        return Responses.list(service.batchCreate(projectId, req));
+    }
+
+    @ApiOperation(value = "batchRevokeTablePermission", notes = "Batch revoke table permissions")
+    @RequestMapping(value = "/batchRevoke", method = RequestMethod.DELETE)
```
Comment (yizhouxw) [inline server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/TablePermissionController.java:92]: use POST method for self defined action

File: server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/SqlAsyncExecuteResp.java [hunk 5]
```diff
@@ -37,7 +37,7 @@ public class SqlAsyncExecuteResp {
     private String requestId;
     private List<Rule> violatedRules;
     private List<SqlTuplesWithViolation> sqls;
-    private List<UnauthorizedDatabase> unauthorizedDatabases;
+    private List<UnauthorizedDBResource> unauthorizedDBResource;
```
Comment (yizhouxw) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/session/model/SqlAsyncExecuteResp.java:40]: unauthorizedDBResources

**Review by yizhouxw — COMMENTED** [review]
we have DBTableController for list tables, what is the expected behaivior for DBTableController after current change?

**Review by MarkPotato777 — APPROVED** [review]
LGTM, great job 👍🏻 @ISADBA @smallsheeeep

**Review by MarkPotato777 — DISMISSED** [review]

**Review by LuckyPickleZZ — DISMISSED** [review]
👍🏿 LGTM

**Comment by smallsheeeep** [comment]
> we have DBTableController for list tables, what is the expected behaivior for DBTableController after current change?

The API in `DBTableController` and `DBTableControllerV1` will be kept as before. The **list** API in `DBTableController` may not be used. These API will not cause data change, so no need to add DB permission check.

File: server/odc-server/src/main/resources/log4j2.xml [hunk 6]
```diff
@@ -344,6 +344,55 @@
             </Routes>
         </Routing>
 
+        <!-- 以下为申请数据表权限任务日志配置 -->
+        <Routing name="AllRoutingApplyTableAppender">
+            <Routes pattern="$${ctx:taskId}">
+                <Route>
+                    <RollingFile name="Rolling-${ctx:taskId}"
+                                 fileName="${LOG_DIRECTORY}/apply-table/${ctx:taskWorkSpace}/${ctx:taskId}/apply-table-task.all"
```
[the remark this answers is not in the document]
Comment (smallsheeeep) [reply server/odc-server/src/main/resources/log4j2.xml:352]: Use ${ctx:taskWorkSpace} to generate different log output filepath for different user. Referring to `TaskContextHolder.trace()`.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/project/ProjectService.java [hunk 7]
```diff
@@ -471,6 +477,15 @@ private void deleteMemberRelatedDatabasePermissions(@NonNull Long userId, @NonNu
         }
     }
 
+    private void deleteMemberRelatedTablePermissions(@NonNull Long userId, @NonNull Long projectId) {
+        List<Long> permissionIds = userTablePermissionRepository.findByUserIdAndProjectId(userId, projectId).stream()
+                .map(UserTablePermissionEntity::getId).collect(Collectors.toList());
+        if (!CollectionUtils.isEmpty(permissionIds)) {
```
[the remark this answers is not in the document]
Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/project/ProjectService.java:483]: Okay! You may have code fetishes, just like me 😂.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowInstanceService.java [hunk 8]
```diff
@@ -734,6 +758,29 @@ private void checkCreateFlowInstancePermission(CreateFlowInstanceReq req) {
         if (authenticationFacade.currentUser().getOrganizationType() == OrganizationType.INDIVIDUAL) {
             return;
         }
+        if (req.getTaskType() == TaskType.EXPORT) {
```
[the remark this answers is not in the document]
Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowInstanceService.java:761]: Yes we can, but it seems not necessary. Can we just keep these codes here?

**Review by yhilmare — APPROVED** [review]
lgtm

**Review by LuckyPickleZZ — APPROVED** [review]

**Review by PeachThinking — APPROVED** [review]
LGTM 👍

**Review by krihy — APPROVED** [review]
lgtm

**Review by JessieWuJiexi — APPROVED** [review]

**Review by yizhouxw — APPROVED** [review]

**Review by MarkPotato777 — APPROVED** [review]
LGTM

**Pull request merged**